### PR TITLE
Manifest API changes necessary for SE-0226

### PR DESCRIPTION
**Pull request by hartbit**
This PR implements the manifest API changes necessary for implementing the other features of SE-0226. I have several questions though:

1. Is it okay if I name the new tools version `5.3`? I think its too late to get this into the `5.2` tools version.

2. Are we okay with `PackageDependencyDescription`'s new `name` property being optional and having a value of `nil` when parsing with tool versions lower than 5.3? This seems to make the most sense, but it hides the fact that the value will never be optional for manifests parsed with tool versions greater than 5.3.

Question: At what stage do we want to make sure there are no duplicate package names? During manifest loading? When creating the package graph?

This is implemented on top of https://github.com/apple/swift-package-manager/pull/2422, so that PR will need to be merged first.

### CODE REVIEW

**Comment by SDGGiesbrecht** [comment]
🎉

File: Sources/PackageLoading/PackageDescription4Loader.swift [hunk 1]
```diff
@@ -297,11 +303,14 @@ extension PackageDependencyDescription {
         }
 
         let requirement = try Requirement(v4: json.get("requirement"))
+        let url = try fixURL(json.get("url"), requirement: requirement)
+        var name: String? = json.get("name")
+
+        if name == nil && toolsVersion >= .v5_3 {
+            name = URL(string: url)?.deletingPathExtension().lastPathComponent
```
Comment (aciidb0mb3r) [inline Sources/PackageLoading/PackageDescription4Loader.swift]: Maybe make a static method on PackageReference, similar to `computeIdentity(packageURL:)`

File: Tests/PackageLoadingTests/PD5_3LoadingTests.swift [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                ]
+            )
+            """
+
+        do {
+            try loadManifestThrowing(stream.bytes) { manifest in
+                return XCTFail("did not generate eror")
+            }
+        } catch ManifestParseError.invalidManifestFormat(let error, diagnosticFile: _) {
+            XCTAssert(error.contains("error: \'product(name:package:)\' is unavailable: the \'package\' argument is now mandatory"))
```
Comment (aciidb0mb3r) [inline Tests/PackageLoadingTests/PD5_3LoadingTests.swift]: Maybe say "the 'package' argument is mandatory as of tools version 5.3"

File: Sources/PackageDescription4/PackageDependency.swift [hunk 3]
```diff
@@ -26,26 +26,74 @@ extension Package.Dependency {
     ///    .package(url: "https://example.com/example-package.git", from: "1.2.3"),
     ///
     /// - Parameters:
+    ///     - name: The name of the package, or nil to deduce it from the URL.
     ///     - url: The valid Git URL of the package.
     ///     - version: The minimum version requirement.
+    @available(_PackageDescription, obsoleted: 5.3)
```
Comment (aciidb0mb3r) [inline Sources/PackageDescription4/PackageDependency.swift]: Please use tools version 5.2 since that's the next release.

**Comment by hartbit** [comment]
@aciidb0mb3r I resolved all comments.

**Comment by hartbit** [comment]
@swift-ci please smoke test

**Comment by hartbit** [comment]
@swift-ci please smoke test

**Pull request merged**